File: augur-engine/src/assets/cards.py
"""Social card compositing using Pillow."""

from __future__ import annotations

from pathlib import Path

from PIL import Image, ImageDraw, ImageFont

CARD_SIZES = {
    "1x1": (1080, 1080),
    "9x16": (1080, 1920),
    "16x9": (1200, 675),
}


def _load_font(size: int) -> ImageFont.FreeTypeFont | ImageFont.ImageFont:
    try:
        return ImageFont.truetype(
            "/usr/share/fonts/truetype/dejavu/DejaVuSerif.ttf", size
        )
    except OSError:
        return ImageFont.load_default()


def _load_mono_font(size: int) -> ImageFont.FreeTypeFont | ImageFont.ImageFont:
    try:
        return ImageFont.truetype(
            "/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf", size
        )
    except OSError:
        return ImageFont.load_default()


def generate_cards(
    image_path: str,
    headline: str,
    brand_name: str,
    horizon_label: str,
    fictive_date: str,
    accent_color: str,
    output_dir: str,
    file_prefix: str,
) -> list[str]:
    """Generate social sharing cards in all 3 ratios. Returns list of output paths."""
    Path(output_dir).mkdir(parents=True, exist_ok=True)
    paths: list[str] = []

    for ratio, (w, h) in CARD_SIZES.items():
        out_path = str(Path(output_dir) / f"{file_prefix}-{ratio}.webp")
        _generate_card(image_path, headline, brand_name, horizon_label,
                       fictive_date, accent_color, w, h, out_path)
        paths.append(out_path)

    return paths


def _generate_card(
    image_path: str,
    headline: str,
    brand_name: str,
    horizon_label: str,
    fictive_date: str,
    accent_color: str,
    width: int,
    height: int,
    output_path: str,
) -> None:
    """Generate a single social card."""
    # Load and resize source image
    src = Image.open(image_path).convert("RGB")
    src = _cover_resize(src, width, height)
    card = src.copy()
    draw = ImageDraw.Draw(card, "RGBA")

    font_size = round(width * 0.04)
    headline_font_size = round(width * 0.05)
    padding = round(width * 0.06)

    # Semi-transparent overlay at bottom
    overlay_top = round(height * 0.55)
    draw.rectangle(
        [(0, overlay_top), (width, height)],
        fill=(0, 0, 0, 178),
    )

    font = _load_font(font_size)
    headline_font = _load_font(headline_font_size)
    mono_font = _load_mono_font(round(font_size * 0.6))

    # Brand name
    draw.text(
        (padding, round(height * 0.60)),
        f"\u263d {brand_name}",
        fill=(255, 255, 255, 230),
        font=font,
    )

    # Horizon label
    draw.text(
        (padding, round(height * 0.66)),
        f"\u2500\u2500 {horizon_label.upper()} \u2500\u2500",
        fill=accent_color,
        font=_load_font(round(font_size * 0.7)),
    )

    # Headline (truncated)
    max_chars = (width - 2 * padding) // max(1, round(headline_font_size * 0.5))
    display_headline = headline[:max_chars - 3] + "..." if len(headline) > max_chars else headline
    draw.text(
        (padding, round(height * 0.74)),
        display_headline,
        fill=(255, 255, 255),
        font=headline_font,
    )

    # Date
    draw.text(
        (padding, round(height * 0.86)),
        f"Foreseen: {fictive_date}",
        fill=(255, 255, 255, 178),
        font=mono_font,
    )

    # AI disclaimer
    draw.text(
        (padding, round(height * 0.92)),
        "AI-generated speculation",
        fill=(255, 255, 255, 128),
        font=_load_mono_font(round(font_size * 0.5)),
    )

    # Bottom accent bar
    draw.rectangle(
        [(0, height - 4), (width, height)],
        fill=accent_color,
    )

    card.convert("RGB").save(output_path, "WEBP", quality=85)


def _cover_resize(img: Image.Image, target_w: int, target_h: int) -> Image.Image:
    """Resize image to cover target dimensions, then center-crop."""
    src_ratio = img.width / img.height
    target_ratio = target_w / target_h

    if src_ratio > target_ratio:
        new_h = target_h
        new_w = round(img.width * (target_h / img.height))
    else:
        new_w = target_w
        new_h = round(img.height * (target_w / img.width))

    img = img.resize((new_w, new_h), Image.LANCZOS)

    left = (new_w - target_w) // 2
    top = (new_h - target_h) // 2
    return img.crop((left, top, left + target_w, top + target_h))

File: augur-engine/src/assets/cards.ts
import sharp from 'sharp';
import { mkdir } from 'node:fs/promises';
import { join, dirname } from 'node:path';

/** Social card aspect ratios */
const CARD_SIZES = {
  '1x1': { width: 1080, height: 1080 },   // Instagram feed, Facebook
  '9x16': { width: 1080, height: 1920 },   // Stories, Reels, TikTok
  '16x9': { width: 1200, height: 675 },    // X/Twitter, OpenGraph
} as const;

type CardRatio = keyof typeof CARD_SIZES;

interface CardOptions {
  imagePath: string;       // source article image
  headline: string;
  brandName: string;       // e.g. "THE AUGUR"
  horizonLabel: string;    // e.g. "TOMORROW"
  fictiveDate: string;
  accentColor: string;     // brand accent hex
  outputDir: string;       // directory for card files
  filePrefix: string;      // e.g. "the-tomorrow-2026-03-10"
}

/** Generate social sharing cards in all 3 ratios */
export async function generateCards(options: CardOptions): Promise<string[]> {
  await mkdir(options.outputDir, { recursive: true });
  const paths: string[] = [];

  for (const [ratio, size] of Object.entries(CARD_SIZES)) {
    const outputPath = join(options.outputDir, `${options.filePrefix}-${ratio}.webp`);
    await generateCard(options, size, outputPath);
    paths.push(outputPath);
    console.log(`[cards] generated ${ratio}: ${outputPath}`);
  }

  return paths;
}

async function generateCard(
  options: CardOptions,
  size: { width: number; height: number },
  outputPath: string,
): Promise<void> {
  const { width, height } = size;
  const fontSize = Math.round(width * 0.04);
  const headlineFontSize = Math.round(width * 0.05);
  const padding = Math.round(width * 0.06);

  // Truncate headline to fit
  const maxHeadlineChars = Math.floor((width - 2 * padding) / (headlineFontSize * 0.5));
  const headline = options.headline.length > maxHeadlineChars
    ? options.headline.slice(0, maxHeadlineChars - 3) + '...'
    : options.headline;

  // Create text overlay SVG
  const svg = `<svg width="${width}" height="${height}">
    <!-- Semi-transparent overlay at bottom -->
    <rect y="${height * 0.55}" width="100%" height="${height * 0.45}" fill="rgba(0,0,0,0.7)"/>

    <!-- Brand name -->
    <text x="${padding}" y="${height * 0.62}"
      font-family="serif" font-weight="bold" font-size="${fontSize}"
      fill="white" letter-spacing="4" opacity="0.9">
      ☽ ${escapeXml(options.brandName)}
    </text>

    <!-- Horizon -->
    <text x="${padding}" y="${height * 0.67}"
      font-family="serif" font-size="${fontSize * 0.7}"
      fill="${options.accentColor}" letter-spacing="3">
      ── ${escapeXml(options.horizonLabel)} ──
    </text>

    <!-- Headline -->
    <text x="${padding}" y="${height * 0.76}"
      font-family="serif" font-weight="bold" font-size="${headlineFontSize}"
      fill="white">
      ${escapeXml(headline)}
    </text>

    <!-- Date + source count -->
    <text x="${padding}" y="${height * 0.87}"
      font-family="monospace" font-size="${fontSize * 0.6}"
      fill="rgba(255,255,255,0.7)">
      🔮 Foreseen: ${escapeXml(options.fictiveDate)}
    </text>

    <!-- AI disclaimer -->
    <text x="${padding}" y="${height * 0.93}"
      font-family="monospace" font-size="${fontSize * 0.5}"
      fill="rgba(255,255,255,0.5)" letter-spacing="1">
      ⚠ AI-generated speculation
    </text>

    <!-- Bottom watermark pattern -->
    <rect y="${height - 4}" width="100%" height="4" fill="${options.accentColor}" opacity="0.6"/>
  </svg>`;

  // Resize source image, composite text overlay
  await sharp(options.imagePath)
    .resize(width, height, { fit: 'cover' })
    .composite([{
      input: Buffer.from(svg),
      top: 0,
      left: 0,
    }])
    .webp({ quality: 85 })
    .toFile(outputPath);
}

function escapeXml(str: string): string {
  return str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}
